File: _backend/pkg/messages/performance/performance.go
package performance

import (
	"math"
)

func TimeDiff(t1 uint64, t2 uint64) uint64 {
	if t1 < t2 {
		return 0
	}
	return t1 - t2
}

func FrameRate(frames int64, dt uint64) float64 {
	return float64(frames) * 1000 / float64(dt)
}

func TickRate(ticks int64, dt uint64) float64 {
	tickRate := float64(ticks) * 30 / float64(dt)
	if tickRate > 1 {
		tickRate = 1
	}
	return tickRate
}

func CPURateFromTickRate(tickRate float64) uint64 {
	return 100 - uint64(math.Round(tickRate*100))
}

func CPURate(ticks int64, dt uint64) uint64 {
	return CPURateFromTickRate(TickRate(ticks, dt))
}

File: tracker/tracker/src/webworker/detectors/performance.ts
/**
 * Port of backend/pkg/messages/performance/performance.go.
 * Used by the CPU issue detector to turn raw frame ticks into a CPU-load %.
 */

export function timeDiff(t1: number, t2: number): number {
  if (t1 < t2) {
    return 0
  }
  return t1 - t2
}

export function tickRate(ticks: number, dt: number): number {
  if (dt === 0) {
    // dt of 0 would divide by zero; treat as a full tick rate (0% cpu load).
    return 1
  }
  let rate = (ticks * 30) / dt
  if (rate > 1) {
    rate = 1
  }
  return rate
}

export function cpuRateFromTickRate(rate: number): number {
  return 100 - Math.round(rate * 100)
}

export function cpuRate(ticks: number, dt: number): number {
  return cpuRateFromTickRate(tickRate(ticks, dt))
}
